cmd/frontend/internal/cli/config_test.go

```diff
@@ -165,11 +167,181 @@ func TestReadSiteConfigFile(t *testing.T) {
 	}
 }
 
+func TestGitserverAddr(t *testing.T) {
+	cases := []struct {
+		name       string
+		deployType string
+		environ    []string
+		want       string
+	}{{
+		name: "test default",
+		want: "gitserver:3178",
+	}, {
+		name:    "default",
+		environ: []string{"SRC_GIT_SERVERS=k8s+rpc://gitserver:3178?kind=sts"},
+		want:    "k8s+rpc://gitserver:3178?kind=sts",
+	}, {
+		name:    "exact",
+		environ: []string{"SRC_GIT_SERVERS=gitserver-0:3178 gitserver-1:3178"},
+		want:    "gitserver-0:3178 gitserver-1:3178",
+	}, {
+		name: "replicas",
+		environ: []string{
+			"SRC_GIT_SERVERS=2",
+		},
+		want: "gitserver-0.gitserver:3178 gitserver-1.gitserver:3178",
+	}, {
+		name:       "replicas helm",
+		deployType: deploy.Helm,
+		environ: []string{
+			"SRC_GIT_SERVERS=2",
+		},
+		want: "gitserver-0.gitserver:3178 gitserver-1.gitserver:3178",
+	}, {
+		name:       "replicas docker-compose",
+		deployType: deploy.DockerCompose,
+		environ: []string{
+			"SRC_GIT_SERVERS=2",
+		},
+		want: "gitserver-0:3178 gitserver-1:3178",
+	}, {
+		name:       "unsupported deploy type",
+		deployType: deploy.PureDocker,
+		environ: []string{
+			"SRC_GIT_SERVERS=5",
+		},
+		want: "",
+	}, {
+		name: "unset",
+		environ: []string{
+			"SRC_GIT_SERVERS=",
+		},
+		want: "",
+	}}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			deploy.Mock(tc.deployType)
+			got, _ := gitserverAddr(tc.environ)
```

Comment:
we should assert on errors as well. A reasonable pattern here is just checking if err contains a subset of wantErr. Sketch code:
```suggestion
			got, err := gitserverAddr(tc.environ)
			if (tc.wantErr == "") != (err == nil) {
				t.Errorf("unexpected error. want=%q got=%v", tc.wantErr, err)
			} else if err != nil && !strings.Contains(err.Error(), tc.wantErr) {
				t.Errorf("error does not contain %q: %v", tc.wantErr, err)
			}
```